src/toolkits/drawing_classifier/drawing_classifier.cpp

```diff
@@ -760,65 +760,39 @@ gl_sframe drawing_classifier::predict_topk(gl_sframe data,
-  size_t topk_class_index_column_idx = dc_predictions.column_index(
-    "topk_class_index");
-  auto get_class_name = [=](const sframe_rows::row& row) {
-    const flex_list& topk_class_index_list = row[topk_class_index_column_idx];
-    flex_list topk_class;
-    for (flexible_type class_index : topk_class_index_list) {
-      topk_class.push_back(class_labels[class_index]);
-    }
-    return topk_class;
-  };
+  auto compute_result_column = [=](const flexible_type& ft) {
+    const flex_vec& prob_vec = ft.get<flex_vec>();
 
-  dc_predictions.add_column(
-      dc_predictions.apply(get_class_name, flex_type_enum::LIST), "class");
+    const flex_list& top_k_indices = argsort_top_k_indices(ft);
 
-  auto generate_ranks = [=](const sframe_rows::row& row) {
-    flex_list ranks(k);
-    std::iota(ranks.begin(), ranks.end(), 0);
-    return ranks;
-  };
+    flex_dict result;
 
-  dc_predictions.add_column(
-      dc_predictions.apply(generate_ranks, flex_type_enum::LIST), "rank");
+    for (size_t rank = 0; rank < k; rank++) {
+      flex_int class_index = top_k_indices[rank];
+      flexible_type value;
 
-  // if output_type is "probability" then change rank to probabilty
-  if (output_type == "probability") {
-    size_t prob_column_index = dc_predictions.column_index("preds");
-    auto get_probability = [=](const sframe_rows::row& row) {
-      const flex_list& topk_class_indices = row[topk_class_index_column_idx];
-      flex_list topk_prob;
-      for (const flexible_type ii : topk_class_indices) {
-        topk_prob.push_back(row[prob_column_index][ii]);
+      if (output_type == "probability") {
+        value = prob_vec[class_index];
+      } else {
+        value = rank;
       }
-      return topk_prob;
-    };
 
-    dc_predictions["rank"] =
-        dc_predictions.apply(get_probability, flex_type_enum::LIST);
-  }
-
-  // construct the final result
-  gl_sframe result = gl_sframe({{"class", dc_predictions["class"]}});
-  result = result.add_row_number();
-
-  // stack data into a column with single element for each row
-  // stack the labels first
-  result = result.stack("class", "class");
+      result.emplace_back(class_labels[class_index], value);
+    }
+    return result;
+  };
 
-  // stack the rank column,
-  gl_sframe stacked_rank =
-      gl_sframe({{"rank", dc_predictions["rank"]}}).stack("rank", "rank");
-  ASSERT_EQ(stacked_rank.size(), result.size());
+  gl_sframe result = gl_sframe();
 
-  result.add_column(stacked_rank["rank"], "rank");
+  // Get a result column that we can stack, from the probability vector
+  result.add_column(dc_predictions["preds"].apply(compute_result_column,
+                                                  flex_type_enum::DICT),
+                    "result");
+  result = result.add_row_number();
+  std::vector<std::string> new_column_names;
+  new_column_names.push_back("class");
+  new_column_names.push_back("rank");
+  result = result.stack("result", new_column_names);
```

Comment:
Could this just be
```
result = result.stack("result", { "class", "rank" });
```
?

Reply:
@nickjong I tried that initially, but I got the following strange compile error:

```
/Users/schhabra/workspace/turicreate/src/toolkits/drawing_classifier/drawing_classifier.cpp:792:19: error: call to member function 'stack' is ambiguous
  result = result.stack("result", {"class", "rank"});
           ~~~~~~~^~~~~
/Users/schhabra/workspace/turicreate/src/core/data/sframe/gl_sframe.hpp:2310:13: note: candidate function
  gl_sframe stack(const std::string& column_name,
            ^
/Users/schhabra/workspace/turicreate/src/core/data/sframe/gl_sframe.hpp:2383:13: note: candidate function
  gl_sframe stack(const std::string& column_name,
            ^
```

This was odd since both candidate functions have different signatures. One takes in a `std::string& new_column_name` to stack a column of dtype array, whereas the other takes in `std::vector<std::string>& new_column_names` to stack a column of dtype dictionary.

Reply:
Weird. You should at least be able to do:
```
result = result.stack("result", std::vector<std::string>{"class", "rank"});
```
(Helping the compiler out by specifying the types more explicitly)